core/src/main/java/org/apache/iceberg/FilteredManifest.java

```diff
@@ -95,60 +94,32 @@ public FilteredManifest caseSensitive(boolean isCaseSensitive) {
+      boolean requireStatsProjection = requireStatsProjection();
+      Collection<String> projectColumns = requireStatsProjection ? withStatsColumns(columns) : columns;
+
+      return CloseableIterable.filter(
+          reader.entries(projection(fileSchema, projectColumns, caseSensitive)),
           entry -> entry != null &&
               evaluator.eval(entry.file().partition()) &&
               metricsEvaluator.eval(entry.file()));
-
     } else {
       return reader.entries(projection(fileSchema, columns, caseSensitive));
     }
   }
 
   CloseableIterable<ManifestEntry> liveEntries() {
-    if ((rowFilter != null && rowFilter != Expressions.alwaysTrue()) ||
-        (partFilter != null && partFilter != Expressions.alwaysTrue())) {
-      Evaluator evaluator = evaluator();
-      InclusiveMetricsEvaluator metricsEvaluator = metricsEvaluator();
-
-      return CloseableIterable.filter(reader.entries(projection(fileSchema, columns, caseSensitive)),
-          entry -> entry != null &&
-              entry.status() != Status.DELETED &&
-              evaluator.eval(entry.file().partition()) &&
-              metricsEvaluator.eval(entry.file()));
-
-    } else {
-      return CloseableIterable.filter(reader.entries(projection(fileSchema, columns, caseSensitive)),
-          entry -> entry != null && entry.status() != Status.DELETED);
-    }
+    return CloseableIterable.filter(allEntries(), entry -> entry != null && entry.status() != Status.DELETED);
   }
 
+  /**
+   * @return an Iterator of DataFile. Makes defensive copies of files before returning
+   */
   @Override
   public Iterator<DataFile> iterator() {
-    if ((rowFilter != null && rowFilter != Expressions.alwaysTrue()) ||
-        (partFilter != null && partFilter != Expressions.alwaysTrue())) {
-      Evaluator evaluator = evaluator();
-      InclusiveMetricsEvaluator metricsEvaluator = metricsEvaluator();
-
-      // ensure stats columns are present for metrics evaluation
-      List<String> projectColumns = Lists.newArrayList(columns);
-      projectColumns.addAll(STATS_COLUMNS); // order doesn't matter
-
-      // if no stats columns were projected, drop them
-      boolean dropStats = Sets.intersection(Sets.newHashSet(columns), STATS_COLUMNS).isEmpty();
-
-      return Iterators.transform(
-          Iterators.filter(reader.iterator(partFilter, projection(fileSchema, projectColumns, caseSensitive)),
-              input -> input != null &&
-                  evaluator.eval(input.partition()) &&
-                  metricsEvaluator.eval(input)),
-          dropStats ? DataFile::copyWithoutStats : DataFile::copy);
-
-    } else {
-      return Iterators.transform(
-          reader.iterator(partFilter, projection(fileSchema, columns, caseSensitive)),
-          DataFile::copy);
-    }
+    boolean dropStats = dropStats();
+    return CloseableIterable.transform(liveEntries(),
+        e -> dropStats ? e.file().copyWithoutStats() : e.file().copy()).iterator();
```

Comment:
Minor: It would be better to move the `dropStats` check outside of the lambda. There's no need to check it inside when we could just use a different lambda that doesn't put a conditional inside a tight loop:

```java
    if (dropStats()) {
      return CloseableIterable.transform(liveEntries(), e -> e.file().copyWithoutStats()).iterator();
    } else {
      return CloseableIterable.transform(liveEntries(), e -> e.file().copy()).iterator();
    }
```

Reply:
Makes sense